src/lib/gssapi/krb5/s4u_gss_glue.c

```diff
@@ -48,37 +48,44 @@ kg_impersonate_name(OM_uint32 *minor_status,
     OM_uint32 major_status;
     krb5_error_code code;
     krb5_creds in_creds, *out_creds = NULL;
+    krb5_data *subject_cert = NULL;
 
     *output_cred = NULL;
     memset(&in_creds, 0, sizeof(in_creds));
 
-    in_creds.client = user->princ;
+    if (user->princ->type == KRB5_NT_X500_PRINCIPAL)
+        subject_cert = user->princ->data;
+    else
+        in_creds.client = user->princ;
     in_creds.server = impersonator_cred->name->princ;
 
     if (impersonator_cred->req_enctypes != NULL)
         in_creds.keyblock.enctype = impersonator_cred->req_enctypes[0];
 
-    k5_mutex_lock(&user->lock);
-
-    if (user->ad_context != NULL) {
-        code = krb5_authdata_export_authdata(context,
-                                             user->ad_context,
-                                             AD_USAGE_TGS_REQ,
-                                             &in_creds.authdata);
-        if (code != 0) {
-            k5_mutex_unlock(&user->lock);
-            *minor_status = code;
-            return GSS_S_FAILURE;
+    /* Client authroization data can only be sent in the request if a user name
+     * is provided */
+    if (user != NULL) {
```

Comment:
I don't think we need this part, it was a limitation in the old approach.
I think we can reset to my last minimal commit, then think what we can improve, why etc.

Reply:
Fixed